src/main/java/io/cdap/plugin/snowflake/sink/batch/StructuredRecordToCSVRecordTransformer.java

```diff
@@ -0,0 +1,137 @@
+
+  public CSVRecord transform(StructuredRecord record) throws IOException {
+    List<String> fieldNames = new ArrayList<>();
+    List<String> values = new ArrayList<>();
+
+    for (Schema.Field field : record.getSchema().getFields()) {
+      String fieldName = field.getName();
+      String value = convertSchemaFieldToString(record.get(fieldName), field, record);
+
+      fieldNames.add(fieldName);
+      values.add(value);
+    }
+
+    return new CSVRecord(fieldNames, values);
+  }
+
+  /**
+   * Convert a schema field to String which can be read by Snowflake.
+   *
+   * @param value field value
+   * @param field schema field
+   * @param record
+   * @return string representing the value in format, which can be understood by Snowflake
+   */
+  public static String convertSchemaFieldToString(Object value, Schema.Field field, StructuredRecord record)
+    throws IOException {
+    // don't convert null to avoid NPE
+    if (value == null) {
+      return null;
+    }
+
+    Schema fieldSchema = field.getSchema();
+
+    if (fieldSchema.isNullable()) {
+      fieldSchema = fieldSchema.getNonNullable();
+    }
+
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+    if (fieldSchema.getLogicalType() != null) {
+      Instant instant;
+      switch (logicalType) {
+        case DATE:
+          // convert epoch day to yyyy-mm-dd format
+          return LocalDate.ofEpochDay((Integer) value).toString();
+        case TIMESTAMP_MICROS:
+          // convert timestamp to ISO 8601 format
+          instant = Instant.ofEpochMilli(TimeUnit.MICROSECONDS.toMillis((Long) value));
+          return instant.toString();
+        case TIME_MICROS:
+          // convert timestamp to HH:mm:ss,SSS
+          instant = Instant.ofEpochMilli(TimeUnit.MICROSECONDS.toMillis((Long) value));
+          return instant.atZone(ZoneOffset.UTC).toLocalTime().toString();
+        case TIMESTAMP_MILLIS:
+          // convert timestamp to ISO 8601 format
+          instant = Instant.ofEpochMilli((Long) value);
+          return instant.toString();
+        case TIME_MILLIS:
+          // convert timestamp to HH:mm:ss,SSS
+          instant = Instant.ofEpochMilli((Long) value);
+          return instant.atZone(ZoneOffset.UTC).toLocalTime().toString();
+        case DECIMAL:
+          return record.getDecimal(field.getName()).toString();
+        default:
+          throw new IllegalArgumentException(
+            String.format("Field '%s' is of unsupported type '%s'", field.getName(), logicalType.getToken()));
+      }
+    }
+
+    switch (fieldSchema.getType()) {
+      // convert to json so it can be saved to Snowflake's variant
+      case RECORD:
+        return StructuredRecordStringConverter.toJsonString((StructuredRecord) value);
+      // convert to json so it can be saved to Snowflake's variant
+      case ARRAY:
+        String stringRecord = StructuredRecordStringConverter.toJsonString(record);
```

Comment:
why are we converting whole record to string?

Reply:
Unfortunately there is no public method to convert just one CDAP field to array. So I had to do this. Since array can have nested arrays records etc in it. This is something I would like to avoid maintaining that code, and possibly causing bugs.

If you could somehow make that method public, it would be super awesome though!

Reply:
Is this correct? Other switch cases are converting field vs this is converting whole record?

Reply:
Yes. A few lines below I retrieve the field I need from json. Not optimal but correct.